src/query/functions/binary/node.go

```diff
@@ -0,0 +1,482 @@
+	}
+}
+
+func (n *binaryNode) processSingleBlock(
+	block block.Block,
+	fn singleScalarFunc,
+) (block.Block, error) {
+	actualFn := actualScalarFunc(fn, n.op.OperatorType, n.op.info.ReturnBool)
+
+	it, err := block.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	builder, err := n.controller.BlockBuilder(it.Meta(), it.SeriesMeta())
+	if err != nil {
+		return nil, err
+	}
+
+	if err := builder.AddCols(it.StepCount()); err != nil {
+		return nil, err
+	}
+
+	for index := 0; it.Next(); index++ {
+		step, err := it.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		values := step.Values()
+		for _, value := range values {
+			builder.AppendValue(index, actualFn(value))
+		}
+	}
+
+	return builder.Build(), nil
+}
+
+// If returnBool is false and op is a comparison function,
+// the function should return the scalar value of lhs if the comparison
+// is true; otherwise, it should return 1 if true, and 0 if false
+func actualFunc(fn mathFunc, op string, returnBool bool) mathFunc {
+	if returnBool || !isComparison(op) {
+		return fn
+	}
+	return func(left, right float64) float64 {
+		// NB (arnikola): safe to check that this equals 1 since this is
+		// only performed on comparison functions where viable values
+		// are only 1 or 0
+		take := fn(left, right) == 1
+		if take {
+			return left
+		}
+		// NB(arnikola): whereas prometheus does not include not matching
+		// points in the output, we here replace them with NaNs instead
+		return math.NaN()
+	}
+}
+
+func (n *binaryNode) processBothSeries(
+	lIter, rIter block.StepIter,
+) (block.Block, error) {
+	lMeta, rMeta := lIter.Meta(), rIter.Meta()
+
+	lSeriesMeta := logical.FlattenMetadata(lMeta, lIter.SeriesMeta())
+	rSeriesMeta := logical.FlattenMetadata(rMeta, rIter.SeriesMeta())
+
+	takeLeft, correspondingRight, lSeriesMeta := intersect(n.op.info.VectorMatching, lSeriesMeta, rSeriesMeta)
+
+	lMeta.Tags, lSeriesMeta = logical.DedupeMetadata(lSeriesMeta)
+
+	// Use metas from only taken left series
+	builder, err := n.controller.BlockBuilder(lMeta, lSeriesMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := builder.AddCols(lIter.StepCount()); err != nil {
+		return nil, err
+	}
+
+	actualFn := actualFunc(n.op.fn, n.op.OperatorType, n.op.info.ReturnBool)
+
+	for index := 0; lIter.Next() && rIter.Next(); index++ {
+		lStep, err := lIter.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		lValues := lStep.Values()
+		rStep, err := rIter.Current()
+		if err != nil {
+			return nil, err
+		}
+
+		rValues := rStep.Values()
+
+		for seriesIdx, lIdx := range takeLeft {
+			rIdx := correspondingRight[seriesIdx]
+			lVal := lValues[lIdx]
+			rVal := rValues[rIdx]
+
+			builder.AppendValue(index, actualFn(lVal, rVal))
+		}
+	}
+
+	return builder.Build(), nil
+}
+
+// computeOrCache figures out if both lhs and rhs are available, if not then it caches the incoming block
+func (n *binaryNode) computeOrCache(ID parser.NodeID, b block.Block) (block.Block, block.Block, error) {
```

Comment:
is this the same as the other place ? If yes, then think about refactor

Reply:
Yeah, this falls under:
TODO: consolidate binary node with logical node, as a lot of behaviour
is shared.

Might do that in a separate commit once this is in. No other binary functions are planned in the near future so shouldn't affect any upcoming commits

Reply:
cool, was just checking